zkevm-circuits/src/evm_circuit/execution/callop.rs

```diff
@@ -54,32 +59,20 @@ pub(crate) struct CallOpGadget<F> {
 }
 
 impl<F: Field> ExecutionGadget<F> for CallOpGadget<F> {
-    const NAME: &'static str = "CALL_STATICCALL";
+    const NAME: &'static str = "CALL_DELEGATECALL_STATICCALL";
 
-    const EXECUTION_STATE: ExecutionState = ExecutionState::CALL_STATICCALL;
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CALL_DELEGATECALL_STATICCALL;
 
     fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
         let opcode = cb.query_cell();
         cb.opcode_lookup(opcode.expr(), 1.expr());
-        let is_call = (OpcodeId::STATICCALL.expr() - opcode.expr())
-            * F::from(OpcodeId::STATICCALL.as_u64() - OpcodeId::CALL.as_u64())
-                .invert()
-                .unwrap();
-
-        // We do the responsible opcode check explicitly here because we're not
-        // using the `SameContextGadget` for `CALL` or `STATICCALL`.
-        cb.require_equal(
-            "Opcode should be CALL or STATICCALL",
-            opcode.expr(),
-            select::expr(
-                is_call.expr(),
-                OpcodeId::CALL.expr(),
-                OpcodeId::STATICCALL.expr(),
-            ),
-        );
+        let is_call = IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::CALL.expr());
+        let is_delegatecall =
+            IsZeroGadget::construct(cb, opcode.expr() - OpcodeId::DELEGATECALL.expr());
+        let is_staticcall = 1.expr() - is_call.expr() - is_delegatecall.expr();
```

Comment:
It looks like we need to use `IsZeroGadget` too. Do we have other checks to prevent this from being something other than STATICCALL? For example, can I witness STOP for the opcode here?

Reply:
Fixed `is_staticcall` to use `IsZeroGadget`. And get the below constraint to verify this opcode must be one of `CALL`, `DELEGATECALL` or `STATICCALL`.
```
cb.require_equal(
    "Opcode should be CALL, DELEGATECALL or STATICCALL",
    is_call.expr() + is_delegatecall.expr() + is_staticcall.expr(),
    1.expr(),
);
```